Add `soft_extended_due_date` field to `Group` model

MattyMay opened the pull request:
Includes server-side validation to ensure `extended_due_date < soft_extended_due_date`

@james-perretta should the student view of a group be changed to only show the `soft_extended_due_date` like how `soft_deadline` and `hard_deadline` work?

MattyMay committed:
Add `soft_extended_due_date` field to `Group` model

Includes server-side validation to ensure `extended_due_date <
soft_extended_due_date`

james-perretta commented:
@MattyMay Yes, students should only see the soft_extended_due_date when requesting their group object. We should chat about how to handle the backwards-incompatibility of the change for API users

MattyMay committed:
Add hard_extended_due_date field, make extended_due_date a property

Migration:
1. Add soft_extended_due_date column
2. Add hard_extended_due_date column
3. Copy extended_due_date column to both new columns
4. Delete extended_due_date column

deprecated extended_due_date behavior:
- reads from soft_extended_due_date field
- writes through to soft_extended_due_date and hard_extended_due_date

MattyMay committed:
Remove hard_extended_due_date from views for non-admins

MattyMay committed:
Add date validation tests for project model and remove debug print

MattyMay committed:
Fix field name in ValidationError msg

MattyMay committed:
Fix broken tests and refactor tests a bit

james-perretta commented on autograder/core/models/group/group.py:
```diff
@@ -93,13 +94,30 @@ def member_names(self) -> List[str]:
 
-    extended_due_date = models.DateTimeField(
+    @property
+    def extended_due_date(self) -> Optional[datetime]:
```
Add docstring:

"""
DEPRECATED. For API legacy use, you can continue to use
this field as usual. Do NOT mix usage of this field and the
two new "hard_extended_due_date" and "soft_extended_due_date" fields.

When this field is set, it indicates that members
of this submission group can submit until this specified
date, overriding the project closing time.
Default value: None
"""

james-perretta commented on autograder/core/tests/test_models/test_group/test_group.py:
```diff
@@ -632,3 +638,66 @@ def test_exception_group_mix_of_student_and_staff(self):
+        hard_extended_due_date = timezone.now() + timezone.timedelta(minutes=5)
+        soft_extended_due_date = hard_extended_due_date + timezone.timedelta(minutes=5)
+
+        with self.assertRaises(exceptions.ValidationError):
```
```python
with self.assertRaises(exceptions.ValidationError) as cm:
    ...

self.assertIn('soft_extended_due_date', cm.exception.message_dict)
```

MattyMay committed:
Add requested changes

- add docstring to extended_deadline
- add exception message check to
  test_error_soft_extended_due_date_after_extended_due_date
- rename test for consistency
